rust/fracpack/src/schema.rs

```diff
@@ -514,3 +517,1703 @@ tuple_impl!(T0 T1 T2 T3 T4 T5 T6 T7 T8 T9 T10 T11 T12);
+        val: &serde_json::Value,
+        dest: &mut Vec<u8>,
+    ) -> Result<(), serde_json::Error> {
+        self.handlers[id].json2frac(schema, ty, val, dest)
+    }
+    fn is_empty_container(&self, id: usize, ty: &CompiledType, value: &serde_json::Value) -> bool {
+        self.handlers[id].is_empty_container(ty, value)
+    }
+}
+
+struct CustomBool;
+
+impl CustomHandler for CustomBool {
+    fn matches(&self, _schema: &CompiledSchema, ty: &CompiledType) -> bool {
+        matches!(
+            ty,
+            CompiledType::Int {
+                bits: 1,
+                is_signed: false
+            }
+        )
+    }
+    fn frac2json(
+        &self,
+        _schema: &CompiledSchema,
+        _ty: &CompiledType,
+        src: &[u8],
+        pos: &mut u32,
+    ) -> Result<serde_json::Value, Error> {
+        Ok(bool::unpack(src, pos)?.into())
+    }
+    fn json2frac(
+        &self,
+        _schema: &CompiledSchema,
+        _ty: &CompiledType,
+        val: &serde_json::Value,
+        dest: &mut Vec<u8>,
+    ) -> Result<(), serde_json::Error> {
+        Ok(bool::deserialize(val)?.pack(dest))
+    }
+    fn is_empty_container(&self, _ty: &CompiledType, _value: &serde_json::Value) -> bool {
+        false
+    }
+}
+
+struct CustomString;
+
+impl CustomHandler for CustomString {
+    fn matches(&self, schema: &CompiledSchema, ty: &CompiledType) -> bool {
+        use CompiledType::*;
+        if let List(item) = ty {
+            matches!(schema.get_by_id(*item), Int { bits: 8, .. })
+        } else {
+            false
+        }
+    }
+    fn frac2json(
+        &self,
+        _schema: &CompiledSchema,
+        _ty: &CompiledType,
+        src: &[u8],
+        pos: &mut u32,
+    ) -> Result<serde_json::Value, Error> {
+        Ok(String::unpack(src, pos)?.into())
+    }
+    fn json2frac(
+        &self,
+        _schema: &CompiledSchema,
+        _ty: &CompiledType,
+        val: &serde_json::Value,
+        dest: &mut Vec<u8>,
+    ) -> Result<(), serde_json::Error> {
+        Ok(<&str>::deserialize(val)?.pack(dest))
+    }
+    fn is_empty_container(&self, _ty: &CompiledType, value: &serde_json::Value) -> bool {
+        value.as_str().map_or(false, |s| s.is_empty())
+    }
+}
+
+struct CustomHex;
+
+impl CustomHandler for CustomHex {
+    fn matches(&self, schema: &CompiledSchema, ty: &CompiledType) -> bool {
+        use CompiledType::*;
+        if let List(item) = ty {
+            !schema.get_by_id(*item).is_variable_size()
+        } else {
+            matches!(ty, FracPack(..)) || !ty.is_variable_size()
+        }
+    }
+    fn frac2json(
+        &self,
+        _schema: &CompiledSchema,
+        ty: &CompiledType,
+        src: &[u8],
+        pos: &mut u32,
+    ) -> Result<serde_json::Value, Error> {
+        let len = if ty.is_variable_size() {
+            u32::unpack(src, pos)?
+        } else {
+            ty.fixed_size()
+        };
+        let start = *pos;
+        let end = *pos + len;
```

Comment:
Is there any risk this could exceed `src.len`?

Reply:
No. The data is validated before we get here.